services/src/main/java/io/grpc/services/BinaryLog.java

```diff
@@ -54,133 +71,197 @@
-  public boolean equals(Object o) {
-    if (!(o instanceof BinaryLog)) {
-      return false;
+    @Override
+    void logTrailingMetadata(Metadata metadata, boolean isServer, byte[] callId) {
+      GrpcLogEntry entry = GrpcLogEntry
+          .newBuilder()
+          .setType(isServer ? Type.SEND_TRAILING_METADATA : Type.RECV_TRAILING_METADATA)
+          .setLogger(isServer ? GrpcLogEntry.Logger.SERVER : GrpcLogEntry.Logger.CLIENT)
+          .setCallId(callIdToProto(callId))
+          .setMetadata(metadataToProto(metadata, maxHeaderBytes))
+          .build();
+      sink.write(entry);
     }
-    BinaryLog that = (BinaryLog) o;
-    return this.maxHeaderBytes == that.maxHeaderBytes
-        && this.maxMessageBytes == that.maxMessageBytes
-        && this.sink.equals(that.sink);
-  }
 
-  @Override
-  public int hashCode() {
-    return Objects.hashCode(maxHeaderBytes, maxMessageBytes);
+    @Override
+    <T> T logOutboundMessage(
+        Marshaller<T> marshaller,
+        T consumeOnce,
+        boolean compressed,
+        boolean isServer,
+        byte[] callId) {
+      ByteBuffer buf = serialize(marshaller, consumeOnce);
+      GrpcLogEntry entry = GrpcLogEntry
+          .newBuilder()
+          .setType(Type.SEND_MESSAGE)
+          .setLogger(isServer ? GrpcLogEntry.Logger.SERVER : GrpcLogEntry.Logger.CLIENT)
+          .setCallId(callIdToProto(callId))
+          .setMessage(messageToProto(buf, compressed, maxMessageBytes))
+          .build();
+      sink.write(entry);
+      return marshaller.parse(toInputStream(buf));
+    }
+
+    @Override
+    <T> T logInboundMessage(
+        Marshaller<T> marshaller,
+        T consumeOnce,
+        boolean compressed,
+        boolean isServer,
+        byte[] callId) {
+      ByteBuffer buf = serialize(marshaller, consumeOnce);
+      GrpcLogEntry entry = GrpcLogEntry
+          .newBuilder()
+          .setType(Type.RECV_MESSAGE)
+          .setLogger(isServer ? GrpcLogEntry.Logger.SERVER : GrpcLogEntry.Logger.CLIENT)
+          .setCallId(callIdToProto(callId))
+          .setMessage(messageToProto(buf, compressed, maxMessageBytes))
+          .build();
+      sink.write(entry);
+      return marshaller.parse(toInputStream(buf));
+    }
+
+    @Override
+    int getMaxHeaderBytes() {
+      return maxHeaderBytes;
+    }
+
+    @Override
+    int getMaxMessageBytes() {
+      return maxMessageBytes;
+    }
+
+    private static <T> ByteBuffer serialize(Marshaller<T> marshaller, T message) {
+      InputStream stream = null;
+      try {
+        stream = marshaller.stream(message);
+        return ByteBuffer.wrap(IoUtils.toByteArray(stream));
+      } catch (IOException e) {
+        // should never happen
+        throw new RuntimeException(e);
+      } finally {
+        GrpcUtil.closeQuietly(stream);
+      }
+    }
+
+    private static InputStream toInputStream(ByteBuffer buffer) {
+      return ReadableBuffers.openStream(ReadableBuffers.wrap(buffer.duplicate()), true);
```

Comment:
This is a nice side effect that I didn't realize. The reason was actually because ByteArrayInputStream does not accept ByteBuffer and I was looking for a way to turn ByteBuffer into an InputStream.